### PR TITLE
feat: add TypedItem to allow for psr/cache:3

### PR DESCRIPTION
**Pull request by bshaffer**
Supports `psr/cache:3.0` and addresses #363 

This PR raises the minimum version supported to PHP 7.1 and adds return types to `SysvCacheItemPool` and `MemoryCacheItemPool`

This change will break BC for anyone extending `SysvCacheItemPool`, but I imagine this is a very unlikely scenario.

### CODE REVIEW

File: tests/BaseTest.php [hunk 1]
```diff
@@ -55,4 +55,13 @@ public function getValidKeyName($key)
     {
         return preg_replace('|[^a-zA-Z0-9_\.! ]|', '', $key);
     }
+
+    protected function getCacheItemClass()
+    {
+        if (\PHP_VERSION_ID >= 80000) {
+            return 'Google\Auth\Cache\TypedItem';
```
Comment (Geolim4) [inline tests/BaseTest.php:62]: You can use this syntax (even if the class does not exists)
```suggestion
            return \Google\Auth\Cache\TypedItem::class;
```

(You can do it with a class import too).

File: src/Cache/TypedItem.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        // We test for two types here due to the fact the DateTimeInterface
+        // was not introduced until PHP 5.5. Checking for the DateTime type as
+        // well allows us to support 5.4.
+        if ($expiration instanceof \DateTimeInterface) {
+            return true;
+        }
+
+        if ($expiration instanceof \DateTime) {
+            return true;
+        }
```
Comment (Geolim4) [inline src/Cache/TypedItem.php:178]:
```suggestion
        if ($expiration instanceof \DateTimeInterface || $expiration instanceof \DateTime) {
            return true;
        }
```

File: src/Cache/TypedItem.php [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    /**
+     * Handles an error.
+     *
+     * @param string $error
+     * @throws \TypeError
+     */
+    private function handleError($error)
+    {
+        if (class_exists('TypeError')) {
```
Comment (Geolim4) [inline src/Cache/TypedItem.php:150]:
```suggestion
        if (class_exists(\TypeError::class)) {
```

File: src/Cache/TypedItem.php [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            throw new \TypeError($error);
+        }
+
+        trigger_error($error, E_USER_ERROR);
```
Comment (Geolim4) [inline src/Cache/TypedItem.php:154]:
```suggestion
        trigger_error($error, \E_USER_ERROR);
```

File: tests/BaseTest.php (hunk 1 above)
Comment (bshaffer) [reply tests/BaseTest.php:62]: I can't because this library currently supports PHP 5.4, which does not support this syntax.

Comment (Geolim4) [reply tests/BaseTest.php:62]: Ohhh my bad, I didn't paid attention to such level backward compatibility.

My apologies.

File: src/Cache/MemoryCacheItemPool.php [hunk 5]
```diff
@@ -55,12 +55,18 @@ public function getItem($key)
...
-            $items[$key] = $this->hasItem($key) ? clone $this->items[$key] : new Item($key);
+        if (\PHP_VERSION_ID >= 80000) {
+            foreach ($keys as $key) {
+                $items[$key] = $this->hasItem($key) ? clone $this->items[$key] : new TypedItem($key);
+            }
+        } else {
+            foreach ($keys as $key) {
+                $items[$key] = $this->hasItem($key) ? clone $this->items[$key] : new Item($key);
+            }
         }
```
Comment (noahdietz) [inline src/Cache/MemoryCacheItemPool.php:64]: I like how explicit it is with two distinct branches, but could it be collapsed into one loop with a variable "false" condition type?

```
$default = \PHP_VERSION_ID >= 80000 ? TypedItem:class : Item:class;
for ($keys as $key) {
  $items[$key] = $this->hasItem($key) ? clone $this->items[$key] : new $default($key);
}
```

File: src/Cache/SysVCacheItemPool.php [hunk 6]
```diff
@@ -99,22 +99,30 @@ public function getItem($key)
...
+                $items[$key] = $this->hasItem($key) ?
+                    clone $this->items[$key] :
+                    new TypedItem($key);
+            }
+        } else {
+            foreach ($keys as $key) {
+                $items[$key] = $this->hasItem($key) ?
+                    clone $this->items[$key] :
+                    new Item($key);
+            }
```
Comment (noahdietz) [inline src/Cache/SysVCacheItemPool.php:117]: Same comment here as above.

**Review by noahdietz — COMMENTED** [review]
_Reviewed from my phone lol_

File: src/Cache/SysVCacheItemPool.php (hunk 6 above)
Comment (bshaffer) [reply src/Cache/SysVCacheItemPool.php]: Done!

File: src/Cache/MemoryCacheItemPool.php (hunk 5 above)
Comment (bshaffer) [reply src/Cache/MemoryCacheItemPool.php]: Done!

File: src/Cache/MemoryCacheItemPool.php [hunk 7]
```diff
@@ -55,12 +55,12 @@ public function getItem($key)
      *   key is not found. However, if no keys are specified then an empty
      *   traversable MUST be returned instead.
      */
-    public function getItems(array $keys = [])
+    public function getItems(array $keys = []): iterable
```
Comment (noahdietz) [inline src/Cache/MemoryCacheItemPool.php:58]: Does the PHP Doc `@return` need to be updated? I'm not sure if there is a difference between `array` and `iterable` here.

**Review by noahdietz — APPROVED** [review]
I will Approve, but perhaps double check with @dwsupplee that the BC is OK? Feel like I'm still learning a bit in these repos.

File: src/Cache/TypedItem.php [hunk 8]
```diff
@@ -0,0 +1,187 @@
+<?php
+/*
+ * Copyright 2016 Google Inc.
```
Comment (dwsupplee) [inline src/Cache/TypedItem.php:3]:
```suggestion
 * Copyright 2022 Google Inc.
```

File: src/Cache/TypedItem.php [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+ */
+
+namespace Google\Auth\Cache;
+
+use Psr\Cache\CacheItemInterface;
+
+/**
+ * A cache item.
+ */
+final class TypedItem implements CacheItemInterface
```
Comment (dwsupplee) [inline src/Cache/TypedItem.php:25]: It looks like the property declaration for `key` is missing.

**Review by dwsupplee — CHANGES_REQUESTED** [review]
I agree the surface for the BC break is small enough to warrant the support here 👍 , LGTM with a few minor notes.

File: src/Cache/TypedItem.php (hunk 9 above)
Comment (Lustmored) [reply src/Cache/TypedItem.php:25]: I see it is defined in the constructor (using property promotion available since PHP 8.0) 👍

Comment (dwsupplee) [reply src/Cache/TypedItem.php:25]: Good call 👍

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -23,16 +23,13 @@ jobs:
...
-              name: Run PHPUnit Patches
-              run: sh .github/apply-phpunit-patches.sh
             - name: Run Script
               run: vendor/bin/phpunit
     test_lowest:
         runs-on: ubuntu-latest
         strategy:
             matrix:
-                php: [ "5.6", "7.2" ]
+                php: [ "7.1", "7.2" ]
```
[the remark this answers is not in the document]
Comment (bshaffer) [reply .github/workflows/tests.yml:32]: done!

File: src/Cache/TypedItem.php (hunk 8 above)
Comment (bshaffer) [reply src/Cache/TypedItem.php]: done!

**Comment by bshaffer** [comment]
@dwsupplee all done here!

**Review by dwsupplee — APPROVED** [review]

**Comment by Geolim4** [comment]
Alleluia 🥳 

Thank you to all participants.

**Comment by bshaffer** [comment]
This is now released in [v1.20.0](https://github.com/googleapis/google-auth-library-php/releases/tag/v1.20.0)

**Pull request merged**